### DISCUSSION TITLE
2 Channel audio / Stereo audio Not working?

### DISCUSSION THREAD

**Discussion by charlielee206** [pschatzmann/arduino-audio-tools#1836] [Q&A] [answered]
**TL;DR: Stereo doesnt work and I dont know how to enable it.**

I'm trying to send stereo audio wirelessly between two ESP32s via ESP-NOW, but I can't seem to get stereo to work.
For mono, everything works flawlessly, but everything breaks apart as soon as I increase the No. of channels to 2.

When i increase the number of channels to 2, the audio gets crunchy, and gets extremely slowed down. I'm guessing the problem is a mono-stereo mismatch, something inverse to this discussion: https://github.com/pschatzmann/arduino-audio-tools/discussions/1178

I'm assuming the esp32 currently can't read stereo at all, since when I run the example "streams_i2c_serial" and monitor the output via the serial plotter, I get only 1 audio signal. I get 4 signals, 1 responds correctly to the input audio signal, but the other 3 are stuck at 1024. (I don't know if the other 3 values are some read/status/log announcement values that the serial plotter automatically picked up as data values. Anyways,I still find it weird that I only have 1 audio signal even in stereo mode.)

I feel like I have to enable stereo mode somewhere, But I can't seem to find it.

<details>
<summary>Hardware</summary>
<br>

(Input/TX) The serial example is also run here.
Audio Source(MP3 Player) -> ADC -> ESP32
ADC: WM8782 I2S ADC board, set to "Slave mode, 16 bits, 48Khz" via the DIP switches and jumper cables.
Wiring: (ADC -> ESP32)
```TO I2S -> Open
MCLK -> RX0
BCLK -> 14
DATA -> 32
LRCK -> 15
DGND, GND -> GND
VCC -> Vin
```

(Output/RX)
ESP32 -> DAC -> headphone jack and speakers
DAC: Sparkfun MAX98357A breakout board (Also tried with other generic MAX98357A boards, all worked in mono mode while not working in stereo)
Wiring: (DAC -> ESP32)
```LRCLK -> 15
BCLK -> 14
DIN -> 22
GAIN -> Vin
!SD -> Vin
GND -> GND
VDD -> Vin
```
</details>

<details>
<summary>Software</summary>
<br>
Please note that i have tried playing with sample frequency, bit depth, buffer no and size, channels, and all that jazz. nothing seemed to work.<br><br>

Semi-Modified Example code:
```#include "AudioTools.h"

AudioInfo info(24000, 2, 16);
I2SStream i2sStream; // Access I2S as stream
CsvOutput<int32_t> csvStream(Serial);
StreamCopy copier(csvStream, i2sStream); // copy i2sStream to csvStream

// Arduino Setup
void setup(void) {
    Serial.begin(115200);
    AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Info);
    
    auto cfg = i2sStream.defaultConfig(RX_MODE);
    cfg.copyFrom(info);
    cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
    cfg.is_master = true;
     // this module nees a master clock if the ESP32 is master
    cfg.use_apll = false;  // try with yes
    cfg.pin_mck = 3; 
    i2sStream.begin(cfg);

    // make sure that we have the correct channels set up
    csvStream.begin(info);

}

// Arduino loop - copy data
void loop() {
    copier.copy();
}
```
<br>

TX Code:
```
#include "AudioTools.h"
#include "AudioTools/Communication/ESPNowStream.h"

#define USE_AUDIO_LOGGING false

//#define RETURN_TO_MONKE


//THIS IS A WORKING SETTING. DO NOT FEKKIN CHANGE THIS
#ifdef RETURN_TO_MONKE 
  #define SAMPLE_RATE 24000
  #define BIT_DEPTH 16
  #define CHANNEL_NO 1
  #define BUFFER_SIZE 128
  #define BUFFER_COUNT 2
  #define INDICATOR 0

//CHANGE THIS
#else
  #define SAMPLE_RATE 24000 //24000
  #define BIT_DEPTH 16 //16
  #define CHANNEL_NO 2 //1
  #define BUFFER_SIZE 64 //128
  #define BUFFER_COUNT 2 //2
  #define INDICATOR 1 //led to show if test mode is on
#endif


AudioInfo info(SAMPLE_RATE, CHANNEL_NO, BIT_DEPTH); 
I2SStream i2sStream; 
ESPNowStream now; 
StreamCopy copier(now, i2sStream);


const char *peers[] = {"A8:48:FA:0B:93:01"};

void setup() {
  Serial.begin(115200);
  pinMode(2,OUTPUT);
  pinMode(27,OUTPUT);
  digitalWrite(27, 0);
  digitalWrite(2, INDICATOR);
  //AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Warning);

  auto cfg = now.defaultConfig();
  cfg.mac_address = "A8:48:FA:0B:93:02";
  now.begin(cfg);
  now.addPeers(peers);

  auto i2sConfig = i2sStream.defaultConfig(RX_MODE);

  i2sConfig.bits_per_sample = BIT_DEPTH;
  i2sConfig.buffer_size = BUFFER_SIZE;
  i2sConfig.buffer_count = BUFFER_COUNT;
  i2sConfig.channels = CHANNEL_NO; 
  i2sConfig.sample_rate = SAMPLE_RATE;

  i2sConfig.is_master = true; 

  i2sConfig.use_apll = false;  
  i2sConfig.pin_mck = 3; 


  i2sStream.begin(i2sConfig);

  Serial.println("Sender started...");
  digitalWrite(27, 1);
}

void loop() { 
  copier.copy();
}
```

<br>
RX Code:

```#include "AudioTools.h"
#include "AudioTools/Communication/ESPNowStream.h"

#define USE_AUDIO_LOGGING false

//#define RETURN_TO_MONKE

//THIS IS A WORKING SETTING. DO NOT FEKKIN CHANGE THIS
#ifdef RETURN_TO_MONKE 
  #define SAMPLE_RATE 24000
  #define BIT_DEPTH 16
  #define CHANNEL_NO 1
  #define BUFFER_SIZE 128
  #define BUFFER_COUNT 2 //2
  #define INDICATOR 0

//CHANGE THIS
#else
  #define SAMPLE_RATE 24000 //24000
  #define BIT_DEPTH 16 //16
  #define CHANNEL_NO 2 //1
  #define BUFFER_SIZE 256 //128
  #define BUFFER_COUNT 4 //2
  #define INDICATOR 1 //LED to show if test mode is on
#endif


AudioInfo info(SAMPLE_RATE, CHANNEL_NO, BIT_DEPTH);
ESPNowStream now;

I2SStream out; 
StreamCopy copier(out, now);     
const char *peers[] = {"A8:48:FA:0B:93:02"};

void setup() {
  pinMode(27,OUTPUT);
  digitalWrite(27, 0);
  Serial.begin(115200);
  pinMode(2,OUTPUT);
  digitalWrite(2, INDICATOR);
  //AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Warning);

  auto cfg = now.defaultConfig();
  cfg.mac_address = "A8:48:FA:0B:93:01";
  now.begin(cfg);
  now.addPeers(peers);

  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);
  config.copyFrom(info); 

  config.bits_per_sample = BIT_DEPTH;
  config.channels = CHANNEL_NO;
  config.sample_rate = SAMPLE_RATE;
  config.buffer_count = BUFFER_COUNT;
  config.buffer_size = BUFFER_SIZE;


  out.begin(config);

  Serial.println("Receiver started...");
  digitalWrite(27, 1);
}

void loop() { 
  copier.copy();
}
```

</details>

Also, Here is a sliver of the log in debug mode. I can't seem to find any useful info for me.

```
327686,262149
[D] StreamCopy.h : 396 - write: 1024 -> 1024
[I] StreamCopy.h : 163 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[D] StreamCopy.h : 100 - copy 1024 bytes
[D] StreamCopy.h : 243 - available: 3072
[D] StreamCopy.h : 390 - write: 1024
[D] AudioOutput.h : 173 - CsvOutput::write: 1024
131074,131073
```
This is all the log parts that I could find. maybe there's more within the flood of data, but i couldn't find any.

**Comment by pschatzmann** [comment, marked as answer]
Changing to stereo is as simple as defining the channels as 2 instead of 1: You will need to do this on the sending and receiving side!

But be careful this doubles the amount of data to be transmitted in the same time: Try to measure first how many samples (or bytes you can transmit per second) with a MeasuringStream and from there calculate the max sample rate that you can use! 

You can try to optimize the values of the ESP-Now configuration (e.g. the rate)

Using a codec will reduce the amount of transmitted data.
